services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloaderSubscriber.java

```diff
@@ -79,6 +81,8 @@ public class MultipartDownloaderSubscriber implements Subscriber<AsyncResponseTr
      */
     private final Object lock = new Object();
 
+    private final List<CompletableFuture<GetObjectResponse>> getObjectFutures = new ArrayList<>();
```

Comment:
Not thread safe.

Since this may potentially use a lot of memory, is there another way to cancel ongoing futures?

Reply:
Updated to use ConcurrentLinkedQueue, and remove each getObjectFuture in .whenComplete()